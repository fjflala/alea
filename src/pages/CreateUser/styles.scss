.create-user {
  height: 100%;
  background-color: #eee;
  padding: 32px;

  &__cancel-button,
  &__create-button {
    margin-top: 16px;
    display: inline-block;
    vertical-align: middle;
  }
  
  &__cancel-button {
    padding-top: 8px;
    margin-right: 8px;
  }

  &__card {
    margin: 32px 0;
  }
}