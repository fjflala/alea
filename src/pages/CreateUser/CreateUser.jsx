/**
 * Module dependencies
 */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

/**
 * UI Components
 */
import Card from '../../components/Card';
import Input from '../../components/Input';
import Button from '../../components/Button';
import Header from '../../components/Header';

/**
 * Import styles
 */
import './styles.scss';
import { usersCreateUser, usersCleanCreatedUser } from '../../modules/Users/actions';

/**
 * Represents a view of Create user
 */
export const CreateUser = ({
  createUser,
  createdUser,
  cleanCreatedUser,
}) => {
  const [user, setUser] = useState({
    name: '',
    job: '',
  });

  useEffect(() => () => {
    cleanCreatedUser();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const onChangeValue = (field, value) => {
    setUser({
      ...user,
      [field]: value,
    });
  };

  const onSubmit = (event) => {
    event.preventDefault();

    createUser({
      data: user,
    });
  };

  return (
    <React.Fragment>
      <Header />
      <section className="create-user">
        <header>
          <h1>Create User</h1>
        </header>
        <Card className="create-user__card">
          {Object.keys(createdUser).length > 0 && (
          <p className="created-user__message--success">
            The new user had been created successfully
            <br />
            the ID of the new user is:
            {createdUser.id}
            <br />
            <Link to="/users">
              Go to users list
            </Link>
          </p>
          )}
          <form onSubmit={onSubmit}>
            <Input
              type="text"
              placeholder="Name"
              onChange={e => onChangeValue('name', e.target.value)}
              value={user.name}
            />
            <Input
              type="text"
              placeholder="Job"
              onChange={e => onChangeValue('job', e.target.value)}
              value={user.job}
            />
            <Link to="/users" className="ui-button ui-button--secondary create-user__cancel-button">
              Cancel
            </Link>
            <Button className="create-user__create-button" type="submit" modifier="primary">
              Create
            </Button>
          </form>
        </Card>
      </section>
    </React.Fragment>
  );
};

/**
 * PropTypes definitions 
 */
CreateUser.propTypes = {
  createdUser: PropTypes.shape({}),
  createUser: PropTypes.func.isRequired,
  cleanCreatedUser: PropTypes.func.isRequired,
};

CreateUser.defaultProps = {
  createdUser: {},
};

/**
 * Connect to store
 */
const mapStateToProps = state => ({
  createdUser: state.users.createdUser,
});

const mapDispatchToProps = {
  createUser: usersCreateUser,
  cleanCreatedUser: usersCleanCreatedUser,
};

/**
 * Expose component
 */
export default connect(mapStateToProps, mapDispatchToProps)(CreateUser);
