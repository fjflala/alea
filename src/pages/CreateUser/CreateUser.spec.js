import React from 'react';
import renderer from 'react-test-renderer';
import { CreateUser } from './CreateUser';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { mount } from 'enzyme'

describe('<CreateUser />', () => {
  let props;

  beforeEach(() => {
    props = {
      createUser: jest.fn(),
      cleanCreatedUser: jest.fn(),
      createdUser: {}
    };
  });

  it('should correctly render', () => {

    const tree = renderer
      .create(<Router history={createMemoryHistory({route: ['/user/create']})}>
        <CreateUser {...props} />
      </Router>)
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  });

  it('should call register method when click on register button', () => {
    const wrapper = mount(<Router history={createMemoryHistory({route: ['/user/create']})}>
      <CreateUser {...props} />
    </Router>);

    wrapper.find('form').simulate('submit');
    
    expect(props.createUser).toHaveBeenCalled();
  });

  it('should show an error message', () => {
    props.createdUser = {
      id: 1,
      first_name: 'Dan',
      last_name: 'Abramov',
      email: 'dan@abramov.net',
    };
    
    const wrapper = mount(<Router history={createMemoryHistory({route: ['/user/create']})}>
      <CreateUser {...props} />
    </Router>);
    
    expect(wrapper.find('.created-user__message--success').length).toBe(1);
  });
});