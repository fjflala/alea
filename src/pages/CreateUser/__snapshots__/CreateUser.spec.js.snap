// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateUser /> should correctly render 1`] = `
Array [
  <header
    className="ui-header"
  >
    <figure>
      <img
        alt="Alea"
        src="https://www.alea.com/wp-content/themes/alea/library/images/logo-alea.svg"
      />
    </figure>
  </header>,
  <section
    className="create-user"
  >
    <header>
      <h1>
        Create User
      </h1>
    </header>
    <div
      className="ui-card create-user__card"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          className="ui-input "
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Name"
          type="text"
          value=""
        />
        <input
          className="ui-input "
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Job"
          type="text"
          value=""
        />
        <a
          className="ui-button ui-button--secondary create-user__cancel-button"
          href="/users"
          onClick={[Function]}
        >
          Cancel
        </a>
        <button
          className="ui-button ui-button--primary create-user__create-button"
          onClick={[Function]}
          type="submit"
        >
          Create
        </button>
      </form>
    </div>
  </section>,
]
`;
